src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -175,7 +165,29 @@ private void CloseInternal()
         private SecurityStatusPal EncryptData(ReadOnlyMemory<byte> buffer, ref byte[] outBuffer, out int outSize)
         {
             ThrowIfExceptionalOrNotAuthenticated();
-            return _context!.Encrypt(buffer, ref outBuffer, out outSize);
+
+            TaskCompletionSource<bool>? waiter = null;
+            while (true)
+            {
+                if (waiter != null)
+                {
+                    waiter.Task.Wait();
```

Comment:
What happens when EncryptData is being called from an async call path... shouldn't we be awaiting this rather than blocking synchronously?

If for some reason that's not necessary, a TCS is overkill and it can just use a ManualResetEventSlim.  But it seems like a TCS is actually necessary because we want to enable synchronous code paths to block and asynchronous code paths to await.

Reply:
I revoked the logic and added LockAsync to the adapter interface. On the "normal" path we should not hit this e.g. attempt to write/encrypt does not conflict with a renegotiation. 

If it does, we wait for the lock and we try it again. In Sync path, everything should be synchronous. For Async we go via async CompleteLockAndWriteAsync()